table_authenticator.go

```diff
@@ -0,0 +1,140 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+)
+
+// TableAuthenticator is an Authenticator which authenticates based
+// on a fixed username / password and table of nodes.
+type TableAuthenticator struct {
+	config *TableAuthenticatorConfig
+	mu     sync.RWMutex
+}
+
+type TableAuthenticatorConfig struct {
+	// NOTE(sean) username / password is part of the config, as this should eventually be "pluggable" against an auth system
+	Username string
+	Password string
+	Nodes    map[string]*TableAuthenticatorNode
+}
+
+type TableAuthenticatorNode struct {
+	NodeID         string
+	CommissionDate *time.Time
+	RetireDate     *time.Time
+	Public         bool
+}
+
+// NewTableAuthenticator creates a newly initialized TableAuthenticator.
+func NewTableAuthenticator() *TableAuthenticator {
+	return &TableAuthenticator{
+		config: &TableAuthenticatorConfig{
+			Nodes: map[string]*TableAuthenticatorNode{},
+		},
+	}
+}
+
+// UpdateConfig updates the config used for authorization.
+func (a *TableAuthenticator) UpdateConfig(config *TableAuthenticatorConfig) {
+	a.mu.Lock()
+	// TODO(sean) protect against ownership bugs by cloning data
+	a.config = config
+	a.mu.Unlock()
+}
+
+// Authorized returns whether or not the given user is authorized to access the given file.
+func (a *TableAuthenticator) Authorized(f *StorageFile, username, password string, hasAuth bool) bool {
+	// TODO(sean) this implementation only uses a single credential for everything,
+	// as can be seen below. later, we probably want to update this
+	return a.authenticated(username, password, hasAuth) || a.allowed(f)
+}
+
+func (a *TableAuthenticator) authenticated(username, password string, hasAuth bool) bool {
+	return hasAuth && username == a.config.Username && password == a.config.Password
+}
+
+func (m *TableAuthenticator) allowed(f *StorageFile) bool {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if m.config == nil {
+		return false
+	}
+	node, ok := m.config.Nodes[f.NodeID]
+	if !ok {
+		return false
+	}
+	return node.CommissionDate != nil && !f.Timestamp.Before(*node.CommissionDate) && node.Public
```

Comment:
This is where the auth policy we agreed on is implemented.

Reply:
Good point. The assumption with the times is that they are all UTC.

It wouldn't be too hard to make that "data driven" in the sheet so we can include a timezone like `2020-01-01 CST`. In the future we may want to revisit the timestamps anyway to include hours, minutes, etc.

Reply:
this is probably fine for now since we aren't being very precise.  but in the future lets keep this in mind.